File: internal/parser/parser.go
package parser

import (
	"fmt"
	"strings"

	"github.com/nbursa/sentience/internal/types"
)

type Parser struct {
	lexer     *Lexer
	curToken  Token
	peekToken Token
	errors    []string
}

func NewParser(l *Lexer) *Parser {
	p := &Parser{
		lexer:  l,
		errors: []string{},
	}
	p.nextToken()
	p.nextToken()
	return p
}

func (p *Parser) nextToken() {
	p.curToken = p.peekToken
	p.peekToken = p.lexer.NextToken()
}

func (p *Parser) ParseProgram() *types.Program {
	program := &types.Program{}
	program.Statements = []types.Statement{}

	for p.curToken.Type != EOF {
		stmt := p.parseStatement()
		if stmt != nil {
			program.Statements = append(program.Statements, stmt)
		}
		p.nextToken()
	}

	return program
}

func (p *Parser) parseStatement() types.Statement {
	switch p.curToken.Type {
	case AGENT:
		return p.parseAgentStatement()
	case MEM:
		return p.parseMemStatement()
	case ON:
		return p.parseOnInputStatement()
	case REFLECT:
		return p.parseReflectStatement()
	case TRAIN:
		return p.parseTrainStatement()
	case GOAL:
		return p.parseGoalStatement()
	case EMBED:
		return p.parseEmbedStatement()
	case LINK:
		return p.parseLinkStatement()
	case IF:
		return p.parseIfStatement()
	case ENTER:
		return p.parseEnterStatement()
	default:
		return nil
	}
}

func (p *Parser) parseAgentStatement() types.Statement {
	stmt := &types.AgentStatement{}

	p.nextToken() // expect agent name
	if p.curToken.Type != IDENT {
		return nil
	}
	stmt.Name = p.curToken.Literal

	p.nextToken() // expect {
	if p.curToken.Type != LBRACE {
		return nil
	}

	stmt.Body = []types.Statement{}
	p.nextToken()

	for p.curToken.Type != RBRACE && p.curToken.Type != EOF {
		bodyStmt := p.parseStatement()
		if bodyStmt != nil {
			stmt.Body = append(stmt.Body, bodyStmt)
		}
		p.nextToken()
	}

	return stmt
}

func (p *Parser) parseMemStatement() types.Statement {
	stmt := &types.MemStatement{}
	p.nextToken()

	if p.curToken.Type != IDENT {
		return nil
	}
	stmt.Target = p.curToken.Literal
	return stmt
}

func (p *Parser) parseOnInputStatement() types.Statement {
	stmt := &types.OnInputStatement{}

	p.nextToken() // expect input
	if p.curToken.Type != INPUT {
		return nil
	}

	p.nextToken() // expect (
	if p.curToken.Type != LPAREN {
		return nil
	}

	p.nextToken() // expect param
	if p.curToken.Type != IDENT {
		return nil
	}
	stmt.Param = p.curToken.Literal

	p.nextToken() // expect )
	if p.curToken.Type != RPAREN {
		return nil
	}

	p.nextToken() // expect {
	if p.curToken.Type != LBRACE {
		return nil
	}

	stmt.Body = []types.Statement{}
	p.nextToken()

	for p.curToken.Type != RBRACE && p.curToken.Type != EOF {
		bodyStmt := p.parseStatement()
		if bodyStmt != nil {
			stmt.Body = append(stmt.Body, bodyStmt)
		}
		p.nextToken()
	}

	return stmt
}

func (p *Parser) parseReflectAccess() types.Statement {
	fmt.Println("[access] start token:", p.curToken.Type, p.curToken.Literal)

	if p.curToken.Type != MEM {
		fmt.Println("[access] expected MEM, got:", p.curToken.Type)
		return nil
	}

	p.nextToken() // expect DOT
	fmt.Println("[access] next token:", p.curToken.Type, p.curToken.Literal)
	if p.curToken.Type != DOT {
		fmt.Println("[access] expected DOT, got:", p.curToken.Type)
		return nil
	}

	p.nextToken() // expect IDENT (short/long)
	fmt.Println("[access] next token:", p.curToken.Type, p.curToken.Literal)
	if p.curToken.Type != IDENT {
		fmt.Println("[access] expected IDENT, got:", p.curToken.Type)
		return nil
	}
	target := p.curToken.Literal

	p.nextToken() // expect [
	fmt.Println("[access] next token:", p.curToken.Type, p.curToken.Literal)
	if p.curToken.Type != LBRACKET {
		fmt.Println("[access] expected LBRACKET, got:", p.curToken.Type)
		return nil
	}

	p.nextToken() // expect STRING
	fmt.Println("[access] next token:", p.curToken.Type, p.curToken.Literal)
	if p.curToken.Type != STRING {
		fmt.Println("[access] expected STRING, got:", p.curToken.Type)
		return nil
	}
	key := p.curToken.Literal

	p.nextToken() // expect ]
	fmt.Println("[access] next token:", p.curToken.Type, p.curToken.Literal)
	if p.curToken.Type != RBRACKET {
		fmt.Println("[access] expected RBRACKET, got:", p.curToken.Type)
		return nil
	}

	fmt.Println("[access] parsed ReflectAccessStatement:", target, key)

	return &types.ReflectAccessStatement{
		MemTarget: target,
		Key:       key,
	}
}

func (p *Parser) parseReflectStatement() types.Statement {
	stmt := &types.ReflectStatement{}

	p.nextToken() // expect {
	if p.curToken.Type != LBRACE {
		return nil
	}

	stmt.Body = []types.Statement{}
	p.nextToken()

	for p.curToken.Type != RBRACE && p.curToken.Type != EOF {
		fmt.Println("[debug] token in reflect:", p.curToken.Type, p.curToken.Literal)
		if p.curToken.Type == MEM {
			access := p.parseReflectAccess()
			if access != nil {
				stmt.Body = append(stmt.Body, access)
				p.nextToken() // advance after ]
				continue
			}
		}

		// fallback
		body := p.parseStatement()
		if body != nil {
			stmt.Body = append(stmt.Body, body)
		} else {
			p.nextToken()
		}
	}

	return stmt
}

func (p *Parser) parseTrainStatement() types.Statement {
	stmt := &types.TrainStatement{}

	p.nextToken() // expect {
	if p.curToken.Type != LBRACE {
		return nil
	}

	stmt.Body = []types.Statement{}
	p.nextToken()

	for p.curToken.Type != RBRACE && p.curToken.Type != EOF {
		bodyStmt := p.parseStatement()
		if bodyStmt != nil {
			stmt.Body = append(stmt.Body, bodyStmt)
		}
		p.nextToken()
	}

	return stmt
}

func (p *Parser) parseGoalStatement() types.Statement {
	stmt := &types.GoalStatement{}

	p.nextToken() // expect :
	if p.curToken.Type != COLON {
		return nil
	}

	p.nextToken() // expect string
	if p.curToken.Type != STRING {
		return nil
	}

	stmt.Value = p.curToken.Literal
	return stmt
}

func (p *Parser) parseEmbedStatement() types.Statement {
	stmt := &types.EmbedStatement{}

	p.nextToken() // expect source
	if p.curToken.Type != IDENT {
		return nil
	}
	stmt.Source = p.curToken.Literal

	p.nextToken() // expect ->
	if p.curToken.Type != ARROW {
		return nil
	}

	p.nextToken() // expect target (could be mem.short)
	targetParts := []string{}

	if p.curToken.Type == MEM {
		targetParts = append(targetParts, p.curToken.Literal)
		p.nextToken()
		if p.curToken.Type == DOT {
			targetParts = append(targetParts, ".")
			p.nextToken()
		}
	}

	if p.curToken.Type == IDENT {
		targetParts = append(targetParts, p.curToken.Literal)
	}

	stmt.Target = strings.Join(targetParts, "")

	return stmt
}

func (p *Parser) parseLinkStatement() types.Statement {
	stmt := &types.LinkStatement{}

	p.nextToken() // from
	if p.curToken.Type != IDENT {
		return nil
	}
	stmt.From = p.curToken.Literal

	p.nextToken() // <-> arrow
	if p.curToken.Type != LINKARROW {
		return nil
	}

	p.nextToken() // to
	if p.curToken.Type != IDENT {
		return nil
	}
	stmt.To = p.curToken.Literal

	return stmt
}

func (p *Parser) parseIfStatement() types.Statement {
	stmt := &types.IfStatement{}

	p.nextToken() // grab condition start
	condParts := []string{}
	for p.curToken.Type != LBRACE && p.curToken.Type != EOF {
		lit := p.curToken.Literal

		// restore quotes for STRING token
		if p.curToken.Type == STRING {
			lit = `"` + lit + `"`
		}

		condParts = append(condParts, lit)
		p.nextToken()
	}

	stmt.Condition = smartJoin(condParts)

	if p.curToken.Type != LBRACE {
		return nil
	}

	stmt.Body = []types.Statement{}
	p.nextToken()

	for p.curToken.Type != RBRACE && p.curToken.Type != EOF {
		bodyStmt := p.parseStatement()
		if bodyStmt != nil {
			stmt.Body = append(stmt.Body, bodyStmt)
		}
		p.nextToken()
	}

	return stmt
}

func (p *Parser) parseEnterStatement() types.Statement {
	stmt := &types.EnterStatement{}

	p.nextToken()
	if p.curToken.Type != IDENT {
		return nil
	}

	stmt.Target = p.curToken.Literal
	return stmt
}

// === HELPERS ===

func smartJoin(parts []string) string {
	out := ""
	for i, part := range parts {
		if i > 0 && isAlphaNum(parts[i-1]) && isAlphaNum(part) {
			out += " "
		}
		out += part
	}
	return out
}

func isAlphaNum(s string) bool {
	for _, ch := range s {
		if ('a' <= ch && ch <= 'z') || ('A' <= ch && ch <= 'Z') || ('0' <= ch && ch <= '9') {
			return true
		}
	}
	return false
}

File: src/parser.rs
use crate::lexer::{Lexer, Token, TokenType};
use crate::types::{Program, Statement};

pub struct Parser<'a> {
    lexer: &'a mut Lexer<'a>,
    cur_token: Token,
    peek_token: Token,
}

impl<'a> Parser<'a> {
    pub fn new(lexer: &'a mut Lexer<'a>) -> Self {
        let first = lexer.next_token();
        let second = lexer.next_token();
        Parser {
            lexer,
            cur_token: first,
            peek_token: second,
        }
    }

    fn next_token(&mut self) {
        self.cur_token = std::mem::replace(&mut self.peek_token, self.lexer.next_token());
    }

    pub fn parse_program(&mut self) -> Program {
        let mut program = Program {
            statements: Vec::new(),
        };
        while self.cur_token.token_type != TokenType::Eof {
            if let Some(stmt) = self.parse_statement() {
                program.statements.push(stmt);
            }
            self.next_token();
        }
        program
    }

    fn parse_statement(&mut self) -> Option<Statement> {
        match self.cur_token.token_type {
            TokenType::Agent => self.parse_agent(),
            TokenType::Mem => self.parse_mem(),
            TokenType::On => self.parse_on_input(),
            TokenType::Reflect => self.parse_reflect(),
            TokenType::Train => self.parse_train(),
            TokenType::Evolve => self.parse_evolve(),
            TokenType::Goal => self.parse_goal(),
            TokenType::Embed => self.parse_embed(),
            TokenType::If => self.parse_if_context_includes(),
            TokenType::Print => self.parse_print(),
            _ => {
                if self.cur_token.token_type == TokenType::Ident
                    && self.peek_token.token_type == TokenType::Equal
                {
                    let key = self.cur_token.literal.clone();
                    self.next_token();
                    self.next_token();
                    let value = self.cur_token.literal.clone();
                    return Some(Statement::Assignment(key, value));
                }

                Some(Statement::Unknown(self.cur_token.literal.clone()))
            }
        }
    }

    fn parse_agent(&mut self) -> Option<Statement> {
        self.next_token();
        let name = self.cur_token.literal.clone();
        if self.peek_token.token_type != TokenType::LBrace {
            return None;
        }
        self.next_token();
        let mut body = Vec::new();
        self.next_token();
        while self.cur_token.token_type != TokenType::RBrace
            && self.cur_token.token_type != TokenType::Eof
        {
            if let Some(inner) = self.parse_statement() {
                body.push(inner);
            }
            self.next_token();
        }
        Some(Statement::AgentDeclaration { name, body })
    }

    fn parse_mem(&mut self) -> Option<Statement> {
        self.next_token();
        let target = self.cur_token.literal.clone();
        Some(Statement::MemDeclaration { target })
    }

    fn parse_on_input(&mut self) -> Option<Statement> {
        self.next_token();
        if self.cur_token.token_type != TokenType::Input {
            return None;
        }
        self.next_token();
        if self.cur_token.token_type != TokenType::LParen {
            return None;
        }
        self.next_token();
        let param = self.cur_token.literal.clone();
        self.next_token();
        if self.cur_token.token_type != TokenType::RParen {
            return None;
        }
        self.next_token();
        if self.cur_token.token_type != TokenType::LBrace {
            return None;
        }
        let mut body = Vec::new();
        self.next_token();
        while self.cur_token.token_type != TokenType::RBrace
            && self.cur_token.token_type != TokenType::Eof
        {
            if let Some(s) = self.parse_statement() {
                body.push(s);
            }
            self.next_token();
        }
        Some(Statement::OnInput { param, body })
    }

    /// Parse either a full `reflect { ... }` block or a single-line `reflect mem.<target>["<key>"]`.
    fn parse_reflect(&mut self) -> Option<Statement> {
        if self.peek_token.token_type == TokenType::LBrace {
            self.next_token(); // cur_token == LBrace
            self.next_token(); // cur_token == Mem

            if self.cur_token.token_type != TokenType::Mem {
                return None;
            }

            let (mem_target, key) = self.expect_dot_and_bracket()?;
            while self.cur_token.token_type != TokenType::RBrace
                && self.cur_token.token_type != TokenType::Eof
            {
                self.next_token();
            }
            return Some(Statement::Reflect {
                body: vec![Statement::ReflectAccess { mem_target, key }],
            });
        }

        self.next_token();
        if self.cur_token.token_type != TokenType::Mem {
            return None;
        }
        if let Some((mem_target, key)) = self.expect_dot_and_bracket() {
            return Some(Statement::ReflectAccess { mem_target, key });
        }
        None
    }

    fn expect_dot_and_bracket(&mut self) -> Option<(String, String)> {
        self.next_token();
        if self.cur_token.token_type != TokenType::Dot {
            return None;
        }

        self.next_token();
        if self.cur_token.token_type != TokenType::Ident {
            return None;
        }
        let mem_target = self.cur_token.literal.clone();

        self.next_token();
        if self.cur_token.token_type != TokenType::LBracket {
            return None;
        }

        self.next_token();
        if self.cur_token.token_type != TokenType::String {
            return None;
        }
        let key = self.cur_token.literal.clone();

        self.next_token();
        if self.cur_token.token_type != TokenType::RBracket
            && self.cur_token.token_type != TokenType::RBrace
        {
            return None;
        }

        Some((mem_target, key))
    }

    fn parse_train(&mut self) -> Option<Statement> {
        self.next_token();
        if self.cur_token.token_type != TokenType::LBrace {
            return None;
        }
        let mut body = Vec::new();
        self.next_token();
        while self.cur_token.token_type != TokenType::RBrace
            && self.cur_token.token_type != TokenType::Eof
        {
            if let Some(s) = self.parse_statement() {
                body.push(s);
            }
            self.next_token();
        }
        Some(Statement::Train { body })
    }

    fn parse_evolve(&mut self) -> Option<Statement> {
        self.next_token();
        if self.cur_token.token_type != TokenType::LBrace {
            return None;
        }
        let mut body = Vec::new();
        self.next_token();
        while self.cur_token.token_type != TokenType::RBrace
            && self.cur_token.token_type != TokenType::Eof
        {
            if let Some(s) = self.parse_statement() {
                body.push(s);
            }
            self.next_token();
        }
        Some(Statement::Evolve { body })
    }

    fn parse_goal(&mut self) -> Option<Statement> {
        self.next_token();
        if self.cur_token.token_type != TokenType::Colon {
            return None;
        }
        self.next_token();
        if self.cur_token.token_type != TokenType::String {
            return None;
        }
        let value = self.cur_token.literal.clone();
        Some(Statement::Goal(value))
    }

    fn parse_embed(&mut self) -> Option<Statement> {
        self.next_token();
        let source = self.cur_token.literal.clone();
        self.next_token();
        if self.cur_token.token_type != TokenType::Arrow {
            return None;
        }
        self.next_token();
        let mut parts = vec![self.cur_token.literal.clone()];
        self.next_token();
        if self.cur_token.token_type == TokenType::Dot {
            self.next_token();
            parts.push(self.cur_token.literal.clone());
        }
        let target = parts.join(".");
        Some(Statement::Embed { source, target })
    }

    fn parse_if_context_includes(&mut self) -> Option<Statement> {
        self.next_token();
        if self.cur_token.token_type != TokenType::Ident || self.cur_token.literal != "context" {
            return None;
        }
        self.next_token();
        if self.cur_token.token_type != TokenType::Ident || self.cur_token.literal != "includes" {
            return None;
        }
        self.next_token();
        if self.cur_token.token_type != TokenType::LBracket {
            return None;
        }
        let mut values = Vec::new();
        loop {
            self.next_token();
            if self.cur_token.token_type == TokenType::String {
                values.push(self.cur_token.literal.clone());
            } else if self.cur_token.token_type == TokenType::RBracket {
                break;
            } else {
                return None;
            }
        }
        self.next_token();
        if self.cur_token.token_type != TokenType::LBrace {
            return None;
        }
        let mut body = Vec::new();
        self.next_token();
        while self.cur_token.token_type != TokenType::RBrace
            && self.cur_token.token_type != TokenType::Eof
        {
            if let Some(s) = self.parse_statement() {
                body.push(s);
            }
            self.next_token();
        }
        Some(Statement::IfContextIncludes { values, body })
    }

    fn parse_print(&mut self) -> Option<Statement> {
        self.next_token();
        if self.cur_token.token_type != TokenType::String {
            return None;
        }
        let val = self.cur_token.literal.clone();
        Some(Statement::Print(val))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer::Lexer;
    use crate::types::Statement;

    #[test]
    fn parse_simple_agent() {
        let input = r#"
            agent Echo {
              mem short
              goal: "Store and reflect"
              on input(msg) {
                embed msg -> mem.short
                reflect { mem.short["msg"] }
              }
              train {
                print "Training"
              }
            }
        "#;
        let mut lexer = Lexer::new(input);
        let mut parser = Parser::new(&mut lexer);
        let program = parser.parse_program();

        assert_eq!(program.statements.len(), 1);
        match &program.statements[0] {
            Statement::AgentDeclaration { name, body } => {
                assert_eq!(name, "Echo");
                assert!(
                    body.iter().any(|s| {
                        matches!(
                            s,
                            Statement::MemDeclaration { target } if target == "short"
                        )
                    }),
                    "expected MemDeclaration {{ target: \"short\" }}"
                );
                assert!(
                    body.iter()
                        .any(|s| { matches!(s, Statement::Goal(g) if g == "Store and reflect") }),
                    "expected Goal(\"Store and reflect\")"
                );
                assert!(
                    body.iter().any(|s| {
                        matches!(
                            s,
                            Statement::OnInput { param, body: _ } if param == "msg"
                        )
                    }),
                    "expected OnInput {{ param: \"msg\" }}"
                );
                assert!(
                    body.iter()
                        .any(|s| matches!(s, Statement::Train { body: _ })),
                    "expected Train {{ body }}"
                );
            }
            _ => panic!("Expected AgentDeclaration"),
        }
    }
}
